### PR TITLE
Selector

### PR DESCRIPTION
**Pull request by glima**
Quoting the Kconfig help verbatim:

    "It is common to have IoT products with LCD display and buttons (arrows)
    and the need to ask for user input. This is a collection of high-level
    node types to deal with that, the *selector* nodes.

    These node types format a multi-line string to be used by other node
    types doing visual output, such as LCD display (grove/lcd-string), in
    conjuntion with others providing input."

    I did my best to have broad tests, but much more than that was tested by
    hand while trying to make the code robust. It's a PITA to write
    validation to this node, so let's grow that with the time.

### CODE REVIEW

**Comment by glima** [comment]
The ``SELECTED`` input and output ports from the node *spec* are missing, doing that now.

**Comment by barbieri** [comment]
didn't read the PR yet, but the module shouldn't be `selector`, rather a broader scope name that will cover all the input-form elements (boolean, integer...). Need to think of a name, maybe something like `textual-input-form`

**Comment by glima** [comment]
* Gustavo Sverzut Barbieri <notifications@github.com> [2015-10-06 16:15:07 -0700]:

> didn't read the PR yet, but the module shouldn't be `selector`, rather a broader scope name that will cover all the input-form elements (boolean, integer...). Need to think of a name, maybe something like `textual-input-form`
>

I agree on a single module for the selector and input forms, but I
could not think of a reasonable name yet, either. Remember that the
selector itself is not a form, just the others. Maybe ``textual-io``
or something like that?

> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/solettaproject/soletta/pull/863#issuecomment-146030768

-- 
Gustavo Lima Chaves
Intel - Open Source Technology Center

File: src/modules/flow/selector/selector.c [hunk 1]
```diff
@@ -0,0 +1,688 @@
...
+#include "sol-flow/selector.h"
+#include "sol-flow-internal.h"
+
+#include "sol-util.h"
+#include "sol-buffer.h"
+
+#include <errno.h>
+
+#define MIN_SIZE(a, b) (((a) < (b)) ? (a) : (b))
+#define MAX_SIZE(a, b) (((a) > (b)) ? (a) : (b))
```
Comment (edersondisouza) [inline src/modules/flow/selector/selector.c]: These 'min' and 'max' macro are becoming more widespread throughout the code. Maybe unify them on utils someday?

File: src/modules/flow/selector/selector.c [hunk 2]
```diff
@@ -0,0 +1,688 @@
...
+    mdata->value_tag = strstr(mdata->format, VALUE_TAG);
+
+    if (!mdata->value_tag) {
+        SOL_WRN("Bad format, no {value} tag: %s", mdata->format);
+        r = -EINVAL;
+        goto err_format;
+    }
+
+    if (mdata->title_tag > mdata->value_tag) {
+        SOL_WRN("Bad format, {title} tag placet after {value} tag: %s",
```
Comment (edersondisouza) [inline src/modules/flow/selector/selector.c]: s/placet/placed

File: src/modules/flow/selector/selector.c (hunk 1 above)
Comment (glima) [inline src/modules/flow/selector/selector.c]: Yeah, it bothers me too. I'll do that.

File: src/modules/flow/selector/selector.c (hunk 2 above)
Comment (glima) [inline src/modules/flow/selector/selector.c]: Thanks.

File: src/modules/flow/selector/selector.json [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+          "data_type": "string",
+          "description": "Defines the sequency of items to be selected by the user. Each incoming packet is added to an internal list. Use the CLEAR port to start over. The strings here will fill in the '{value}' field of the format option (e. g. '{title}:\\n{value} px'), when sent to output. Any line breaking characters is the input strings for this port will be translated to a space instead.",
+          "methods": {
+            "process": "add_item"
+          },
+          "name": "ADD_ITEM"
+        },
+        {
+          "data_type": "any",
+          "description": "Change selection to the next one in the sequence",
```
Comment (edersondisouza) [inline src/modules/flow/selector/selector.json]: I think that this description is not related to the port CLEAR.

File: src/modules/flow/selector/selector.json [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+          {
+            "data_type": "string",
+            "default": "* ",
+            "description": "If set, selected items will be laid out with this string pre-fixed, separated by a space",
+            "name": "selection_marker"
+          },
+          {
+            "data_type": "string",
+            "default": "> ",
+            "description": "If set, selected items will be laid out with this string pre-fixed, separated by a space",
```
Comment (edersondisouza) [inline src/modules/flow/selector/selector.json]: Do options `cursor_marker` and `selection_marker` have the same descriptions?

File: src/modules/flow/selector/selector.json [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        ],
+        "version": 1
+      },
+      "out_ports": [
+        {
+          "data_type": "string",
+          "description": "The formatted, selected entry. This is meant to feed real display nodes (like grove/lcd-string).",
+          "name": "STRING"
+        }
+      ],
```
Comment (edersondisouza) [inline src/modules/flow/selector/selector.json]: Not sure if I understood node purpose, but shouldn't it output raw value too? To use on the flow?

File: src/modules/flow/selector/selector.c (hunk 1 above)
Comment (ibriano) [inline src/modules/flow/selector/selector.c]: On Wed, 07 Oct 2015 07:23:44 -0700, Gustavo Lima Chaves wrote:
> > + * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
> > + * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
> > + * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
> > + * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
> > + */
> > +
> > +#include "sol-flow/selector.h"
> > +#include "sol-flow-internal.h"
> > +
> > +#include "sol-util.h"
> > +#include "sol-buffer.h"
> > +
> > +#include <errno.h>
> > +
> > +#define MIN_SIZE(a, b) (((a) < (b)) ? (a) : (b))
> > +#define MAX_SIZE(a, b) (((a) > (b)) ? (a) : (b))
> 
> Yeah, it bothers me too. I'll do that.

And since we have compound statements and other things like it all over
the place, might be good to use those to ensure we are comparing the
same type.

File: src/modules/flow/selector/selector.json (hunk 4 above)
Comment (glima) [inline src/modules/flow/selector/selector.json]: Argh, these json docs really need love. Thanks for spotting.

File: src/modules/flow/selector/selector.json (hunk 5 above)
Comment (glima) [inline src/modules/flow/selector/selector.json]: There'll be another output with the raw value of the selection, yes.

File: src/modules/flow/selector/selector.c [hunk 6]
```diff
@@ -0,0 +1,688 @@
...
+        SOL_WRN("Selector columns number must be a positive integer, "
+            "but %d was given.", opts->columns.val);
+        return -EINVAL;
+    } else
+        mdata->columns = opts->columns.val;
+
+    if (opts->rows.val <= 0) {
+        SOL_WRN("Selector rows number must be a positive integer, "
+            "but %d was given.", opts->rows.val);
+        return -EINVAL;
```
Comment (bdilly) [inline src/modules/flow/selector/selector.c]: Just set rows to "1" and continue. Sure, warning should say that.
We're trying to follow this approach in our node types to be less frustrating as possible to users.

File: src/modules/flow/selector/selector.c (hunk 2 above)
Comment (bdilly) [inline src/modules/flow/selector/selector.c]: It doesn't seem to be explained on "format" description.
Also, instead of just aborting here I would suggest to use default format.

File: src/modules/flow/selector/selector.c [hunk 7]
```diff
@@ -0,0 +1,688 @@
...
+        mdata->rows = opts->rows.val;
+
+    mdata->format = strdup(opts->format);
+    SOL_NULL_CHECK(mdata->format, -ENOMEM);
+
+    mdata->title_tag = strstr(mdata->format, TITLE_TAG);
+    mdata->value_tag = strstr(mdata->format, VALUE_TAG);
+
+    if (!mdata->value_tag) {
+        SOL_WRN("Bad format, no {value} tag: %s", mdata->format);
```
Comment (bdilly) [inline src/modules/flow/selector/selector.c]: Can't you just use default format in this case (warning the user about it)?

File: src/modules/flow/selector/selector.c [hunk 8]
```diff
@@ -0,0 +1,688 @@
...
+        goto err_format;
+    }
+
+    mdata->text_mem = calloc(1, get_buf_size(mdata));
+    if (!mdata->text_mem) {
+        r = -ENOMEM;
+        goto err_format;
+    }
+
+    if (opts->title) {
```
Comment (bdilly) [inline src/modules/flow/selector/selector.c]: Shouldn't you check if mdata->title_tag is set here?
Does it makes sense to have a title and don't use it in "format"?
Also, if title_tag is set and no title was defined, will you display "Bla (null): 10" for a format "Bla {title}: {value}"?
I would suggest to warn users.

File: src/modules/flow/selector/selector.c [hunk 9]
```diff
@@ -0,0 +1,688 @@
...
+static void
+selector_close(struct sol_flow_node *node, void *data)
+{
+    struct selector_data *mdata = data;
+    uint16_t i;
+    char *it;
+
+    sol_buffer_fini(&mdata->text_grid);
+
+    free(mdata->format);
```
Comment (bdilly) [inline src/modules/flow/selector/selector.c]: Didn't you forget to free mdata->sel_mark, mdata->title and mdata->text_mem here?
Have you tried to run your samples with valgriind?

**Comment by bdilly** [comment]
@glima  I'm missing tests and samples here, even for string generator.
I remember to see a test in your desk some days ago. Did you forget to commit it?

File: src/modules/flow/selector/selector.c (hunk 6 above)
Comment (glima) [inline src/modules/flow/selector/selector.c]: I thought about that, but is a 1x1 display any more useful than no display at all (error)?

File: src/modules/flow/selector/selector.c (hunk 7 above)
Comment (glima) [inline src/modules/flow/selector/selector.c]: A bit weird to *tratorar* the user format, but yes, I could.

File: src/modules/flow/selector/selector.c (hunk 8 above)
Comment (glima) [inline src/modules/flow/selector/selector.c]: Mmm, I guess it's better to skip this if ``mdata->title_tag`` is NULL, indeed.

File: src/modules/flow/selector/selector.c (hunk 9 above)
Comment (glima) [inline src/modules/flow/selector/selector.c]: Thanks. Weird thing is that valgrind said nothing on ctrl+c.

**Comment by glima** [comment]
Sorry, I forgot to add ``src/test-fbp/selector.fbp``. Doing it now on v2.

**Comment by glima** [comment]
v2 coming, thanks to all.
